### PR TITLE
vulkan-tools 1.3.248 (new formula)

### PR DESCRIPTION
**Pull request by pthariensflame**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

### CODE REVIEW

**Comment by pthariensflame** [comment]
Need help with diagnosing what's going on here.  MoltenVK can't find Metal on the Intel Mac CI machines, but doesn't have that trouble on the Apple Silicon ones?  I'm stumped.

**Comment by pthariensflame** [comment]
And now it's failing on macOS 12 Arm?  😖

**Comment by pthariensflame** [comment]
@carlocab Should be ready now!  I just disabled the MoltenVK part of the test on the Intel runners; I suspect it has to do with not passing through Metal support in the VMs, but don't know how I'd prove that.

File: Formula/vulkan-tools.rb
Comment (carlocab) [inline Formula/vulkan-tools.rb:42]: Are you sure this is used? It doesn't look like you're using it anywhere.

File: Formula/vulkan-tools.rb
Comment (carlocab) [inline Formula/vulkan-tools.rb:56]:
```suggestion
      inreplace buildpath.glob("*/macOS/*/*.cmake") do |s|
        s.gsub! path,
                "${MOLTENVK_DIR}/MoltenVK/include",
                "${MOLTENVK_DIR}/include"
        s.gsub! path,
                "${MOLTENVK_DIR}/MoltenVK/dylib/macOS/libMoltenVK.dylib",
                "${MOLTENVK_DIR}/lib/libMoltenVK.dylib"
      end
```

File: Formula/vulkan-tools.rb
Comment (carlocab) [inline Formula/vulkan-tools.rb:74]:
```suggestion
    args += if OS.mac?
      ["-DMOLTENVK_REPO_ROOT=#{Formula["molten-vk"].prefix}"]
    else
      [
```

File: Formula/vulkan-tools.rb
Comment (carlocab) [inline Formula/vulkan-tools.rb:86]:
```suggestion
                                shared_library("build/icd/libVkICD_mock_icd")
```

File: Formula/vulkan-tools.rb
Comment (carlocab) [inline Formula/vulkan-tools.rb:92]:
```suggestion
    return unless OS.mac?

    bin.install_symlink prefix/"vulkaninfo/vulkaninfo"
    (bin/"vkcube").write_env_script "/usr/bin/open", "-a #{prefix/"cube/vkcube.app"}", {}
    (bin/"vkcubepp").write_env_script "/usr/bin/open", "-a #{prefix/"cube/vkcubepp.app"}", {}
```

File: Formula/vulkan-tools.rb
Comment (carlocab) [inline Formula/vulkan-tools.rb:98]:
```suggestion
        #{opt_prefix}/mock_icd
```

File: Formula/vulkan-tools.rb
Comment (carlocab) [inline Formula/vulkan-tools.rb:103]: Do users need to set this environment variable so that the mock ICD files can be found?

File: Formula/vulkan-tools.rb
Comment (pthariensflame) [reply Formula/vulkan-tools.rb:56]: Applied

File: Formula/vulkan-tools.rb
Comment (pthariensflame) [reply Formula/vulkan-tools.rb:86]: Applied!

File: Formula/vulkan-tools.rb
Comment (pthariensflame) [reply Formula/vulkan-tools.rb:74]: Applied!

File: Formula/vulkan-tools.rb
Comment (pthariensflame) [reply Formula/vulkan-tools.rb:92]: Applied!

File: Formula/vulkan-tools.rb
Comment (pthariensflame) [reply Formula/vulkan-tools.rb:98]: Applied!

File: Formula/vulkan-tools.rb
Comment (pthariensflame) [reply Formula/vulkan-tools.rb:103]: They do, and it shouldn't be set by default because otherwise it can interfere with any actual Vulkan drivers on the platform.  I'll extend the caveat to cover that.

File: Formula/vulkan-tools.rb
Comment (pthariensflame) [reply Formula/vulkan-tools.rb:42]: It appears not!  This was put in because it was a stated requirement in the build docs of the source repo, but if it's not actually needed I'm happy to have it gone!

File: Formula/vulkan-tools.rb
Comment (carlocab) [inline Formula/vulkan-tools.rb:68]:
```suggestion
    (lib/"mock_icd").install (buildpath/"build/icd/VkICD_mock_icd.json").realpath,
                             shared_library("build/icd/libVkICD_mock_icd")
```
Maybe?

File: Formula/vulkan-tools.rb
Comment (carlocab) [inline Formula/vulkan-tools.rb:72]: Why not just install this directly into `bin`?

File: Formula/vulkan-tools.rb
Comment (carlocab) [inline Formula/vulkan-tools.rb:74]: Does this work?
```suggestion
    (bin/"vkcube").write_exec_script prefix/"cube/vkcube.app/Contents/MacOS/vkcube"
    (bin/"vkcubepp").write_exec_script prefix/"cube/vkcubepp.app/Contents/MacOS/vkcubepp"
```

File: Formula/vulkan-tools.rb
Comment (carlocab) [inline Formula/vulkan-tools.rb:80]:
```suggestion
        #{opt_prefix}/mock_icd
```

File: Formula/vulkan-tools.rb
Comment (carlocab) [inline Formula/vulkan-tools.rb:82]:
```suggestion
        export VK_ICD_FILENAMES=#{opt_prefix}/mock_icd/VkICD_mock_icd.json
```

File: Formula/vulkan-tools.rb
Comment (carlocab) [inline Formula/vulkan-tools.rb:74]: Also, I noticed that these `*.app`s have these dylibs:
```
❯ l cube/*/Contents/Frameworks
cube/vkcube.app/Contents/Frameworks:
Permissions Size User     Date Modified Name
.r--r--r--  7.0M carlocab 14 Apr 05:32  libMoltenVK.dylib
.rw-r--r--  534k carlocab 14 Apr 05:32  libvulkan.1.3.247.dylib
lrwxr-xr-x    23 carlocab 14 Apr 05:32  libvulkan.1.dylib -> libvulkan.1.3.247.dylib

cube/vkcubepp.app/Contents/Frameworks:
Permissions Size User     Date Modified Name
.r--r--r--  7.0M carlocab 14 Apr 05:32  libMoltenVK.dylib
.rw-r--r--  534k carlocab 14 Apr 05:32  libvulkan.1.3.247.dylib
lrwxr-xr-x    23 carlocab 14 Apr 05:32  libvulkan.1.dylib -> libvulkan.1.3.247.dylib
```
Have these just been copied from elsewhere?

**Review by carlocab — COMMENTED** [review]
Thanks; almost there!

File: Formula/vulkan-tools.rb
Comment (pthariensflame) [reply Formula/vulkan-tools.rb:68]: Maybe?  There's a tension there because normally you install the JSON file for the ICD in a different place than the dynamic library (share vs. lib) but in this case the JSON insists that the dynamic library be right next to it in the same folder, contradicting the general conventions.  I'll leave it up to you to decide where they should both go.

File: Formula/vulkan-tools.rb
Comment (pthariensflame) [reply Formula/vulkan-tools.rb:74]: Sadly not; the applications fail to start unless you launch them *as* applications, I think because of failing to find the bundled frameworks.

Comment (pthariensflame) [reply Formula/vulkan-tools.rb:74]: Indeed!  And the apps as built expect to find them in those bundled spots.  They're copied from `molten-vk` and `vulkan-loader` so at least it's all within Homebrew still?

File: Formula/vulkan-tools.rb
Comment (pthariensflame) [reply Formula/vulkan-tools.rb:72]: My reasoning was that I didn't want to duplicate the executable any more than necessary, but I can switch that if you'd prefer!

File: Formula/vulkan-tools.rb
Comment (carlocab) [reply Formula/vulkan-tools.rb:74]: Oops, my suggestion was wrong, try this:
```suggestion
    bin.write_exec_script prefix/"cube/vkcube.app/Contents/MacOS/vkcube"
    bin.write_exec_script prefix/"cube/vkcubepp.app/Contents/MacOS/vkcubepp"
```

Comment (pthariensflame) [reply Formula/vulkan-tools.rb:74]: Ah, I didn't even try your suggestion as written because I'd already tried the correct one before.  That's how I arrived at the solution I did.

Comment (carlocab) [reply Formula/vulkan-tools.rb:74]: Ok, that's fine. Let's replace them with symlinks:
```ruby
targets = Formula["molten-vk"].opt_lib.glob(shared_library("*")) +
          Formula["vulkan-loader"].opt_lib.glob(shared_library("*"))
prefix.glob("cube/*.app/Contents/Frameworks").each do |framework_dir|
  ln_sf targets, framework_dir
end
```

File: Formula/vulkan-tools.rb
Comment (carlocab) [reply Formula/vulkan-tools.rb:68]: Actually, let's install `libVkICD_mock_icd` into `lib`, and then `VkICD_mock_icd.json` into `share/"vulkan/icd.d"`. We can `inreplace` the `library_path` to point to the right place.

File: Formula/vulkan-tools.rb
Comment (carlocab) [reply Formula/vulkan-tools.rb:72]: Well if we did
```
bin.install prefix/"vulkaninfo/vulkaninfo"
```
then we'd only have one copy of it.

Comment (pthariensflame) [reply Formula/vulkan-tools.rb:72]: Wouldn't we still have the copy in `#{prefix}/vulkaninfo` since that's where the build system puts it?

Comment (carlocab) [reply Formula/vulkan-tools.rb:72]: No, `bin.install` will move it.

File: Formula/vulkan-tools.rb
Comment (carlocab) [reply Formula/vulkan-tools.rb:74]: Ok, that's fine then!

File: Formula/vulkan-tools.rb
Comment (pthariensflame) [reply Formula/vulkan-tools.rb:68]: I'm very wary of doing that because it might screw up whatever other Vulkan ICD is present on the system…like MoltenVK.  This mock ICD is explicitly only for testing (it's the "null" driver and implements nothing meaningful) and the repo [warns that it should not be installed into default positions](https://github.com/KhronosGroup/Vulkan-Tools/blob/main/BUILD.md):
> The Mock ICD is not installed by default since it is a "null" driver that does not render anything and is used for testing purposes. Installing it to system directories may cause some applications to discover and use this driver instead of other full drivers installed on the system.

Comment (carlocab) [reply Formula/vulkan-tools.rb:68]: Ok, fair enough. `lib/"mock_icd"` it is then.

File: Formula/vulkan-tools.rb
Comment (carlocab) [inline Formula/vulkan-tools.rb:80]:
```suggestion
    (bin/"vkcube").write_env_script "/usr/bin/open", "-a #{prefix}/cube/vkcube.app", {}
    (bin/"vkcubepp").write_env_script "/usr/bin/open", "-a #{prefix}/cube/vkcubepp.app", {}
```

**Review by carlocab — COMMENTED** [review]
Last suggestion; apologies for missing this earlier!

File: Formula/vulkan-tools.rb
Comment (carlocab) [inline Formula/vulkan-tools.rb:68]:
```suggestion
                             shared_library("build/icd/libVkICD_mock_icd")
```

File: Formula/vulkan-tools.rb
Comment (carlocab) [inline Formula/vulkan-tools.rb:69]:
```suggestion
                             shared_library("build/icd/libVkICD_mock_icd")
```

File: Formula/vulkan-tools.rb
Comment (carlocab) [inline Formula/vulkan-tools.rb:74]: Your changes here mean that you don't actually need a `glob` anymore, since you use no glob characters.

Comment (pthariensflame) [reply Formula/vulkan-tools.rb:74]: Oh, okay!

File: Formula/vulkan-tools.rb
Comment (carlocab) [inline Formula/vulkan-tools.rb:74]:
```suggestion
      Formula["vulkan-loader"].opt_lib/shared_library("libvulkan", Formula["vulkan-loader"].version),
```
Might need to do this to make it compatible with Linux; not sure. Using a glob would be simpler.

**Review by carlocab — DISMISSED** [review]
Nice work!

**Comment by pthariensflame** [comment]
(Wanted to refresh the bottles after the #129036 was merged.)

**Review by carlocab — DISMISSED** [review]

**Comment by pthariensflame** [comment]
Anything more needed for merge?

**Comment by pthariensflame** [comment]
@carlocab ❓

**Comment by carlocab** [comment]
Was waiting to see if anyone else had more feedback. But I think we can merge this now.

**Pull request merged**
